### DISCUSSION TITLE
Idea: Allow custom event types to be implmented or a standard event for polling

### DISCUSSION THREAD

**Discussion by Ancient-Dragon** [enisdenjo/graphql-sse#53] [Ideas]
The reason I'm asking about this idea is because we have introduced subscriptions and we have noticed that on our backend the subscription thread stays alive even after the user has closed the browser tab. So I was thinking would it be possible to allow users to add the ability to customise the handling of event types (e.g. if we wanted to implement a poll event then we could add a method to be called on recieving the event). Or if there could be a standard poll event type whereby if this event was recieved nothing would be triggered and it would just kind of be ignored. My assumption here is that the service doesn't know that the client has disconnected because of the unidirecitonal communication, so only when it tries to publish to the connection does it find out it's been closed.

**Comment by enisdenjo** [comment]
graphql-sse already [pings clients every 12 seconds](https://github.com/enisdenjo/graphql-sse/blob/1d230ae7ce728f8df3d3c009ce25b2dfa4be47e0/src/handler.ts#L517-L518).

Maybe 12 seconds is too much for your server, I can add an option to customise the timeout - what do you think?

**Reply by Ancient-Dragon** [reply]
Ah so the thing is we use dgs which is a Java framework, I couldn’t see anything about pinging in the protocol. Would you be able to explain which event type to use or where the documentation is around here so I can update the dgs library to implement the pinging?

**Reply by enisdenjo** [reply]
There's no section about pings in the GraphQL over SSE spec because it is an extension of the SSE spec.

> Legacy proxy servers are known to, in certain cases, drop HTTP connections after a short timeout. To protect against such proxy servers, authors can include a comment line (one starting with a ':' character) every 15 seconds or so.

[_SSE spec "Authoring notes" section_](https://html.spec.whatwg.org/multipage/server-sent-events.html#authoring-notes)

Essentially, issuing an empty comment message (`:\n\n`) is considered a keep-alive ping. The graphql-sse client ignores these empty messages, and so does the browser native `EventSource`.

**Reply by Ancient-Dragon** [reply]
Ah ok thanks. And just to clarify so I get this right, it would be an event type of next with the data being the :\n\n?

**Reply by enisdenjo** [reply]
No, no event type. The message is an empty comment, so it literally looks like this:
```
:


```

A colon `:` followed with 2 empty lines (which denote the end of message).

**Reply by Ancient-Dragon** [reply]
Thanks! I’ll try and test this out later today and see if I can get it working. Appreciate it!

**Reply by Ancient-Dragon** [reply]
After a battle with spring I was able to get it working, thanks for the help! :)

**Reply by enisdenjo** [reply]
Wohoo, glad to hear that!
